### PR TITLE
add `register_batch` and `read_batch` to `SessionContext` to register a single RecordBatch as a table

### LINKED ISSUE #3426: Make a convenience function to register a single `RecordBatch` as a table from SessionContext
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

For testing (and those new to DataFusion) it would be very convenient to run a query against a single `RecordBatch` as a table. This can be done with [MemTable](https://docs.rs/datafusion/11.0.0/datafusion/datasource/memory/struct.MemTable.html) but it adds that much more boiler plate / cognitive overhead to using DataFusion. 

For example the MemTable allows defining multiple "streams" of record batches in multiple partitions (this is what the `Vec<Vec<_>>` is all about)

However, most of the uses of MemTable in DataFusion (see [query](https://github.com/apache/arrow-datafusion/search?q=MemTable)) look like [this](https://github.com/apache/arrow-datafusion/blob/ca5339bfd27677c165263aa01f263c8fba886a45/datafusion/core/tests/dataframe.rs#L66):

```rust
    let table1 = MemTable::try_new(schema1, vec![vec![batch1]])?;
    ctx.register_table("aa", Arc::new(table1))?;
```

https://docs.rs/datafusion/11.0.0/datafusion/execution/context/struct.SessionContext.html#method.register_table

**Describe the solution you'd like**

What I would like is a function on `SessionContext` that does that work

```rust
impl SessionContext {

...
  /// Registers the RecordBatch as the specified table name
  fn register_batch(table_name: &str, batch: RecordBatch) -> Result<Option<Arc<dyn TableProvider>>> {
     // make a memtable here, return it
     todo!()
  }
}
```

And then replace the uses of `MemTable` in the datafusion tests with this new function
**Describe alternatives you've considered**

**Additional context**

I think this is a good first issue as the code already exists and the changes would be largely mechanical: updating the codebase / examples / documentation would be a good exercise for someone to get more experience with it

### PR DESCRIPTION
**Pull request by BaymaxHWY**
# Which issue does this PR close?

Closes #3426

 # Rationale for this change

# What changes are included in this PR?
1. add `SessionContext::register_batch` to register a single `RecordBatch` as a table
2. add `SessionContext::read_batch` to read a `DataFrame` from a single `RecordBatch`
3. replace some code with `register_batch` in test/example

# Are there any user-facing changes?

### CODE REVIEW

**Comment by BaymaxHWY** [comment]
Some of the test code is not a single `RecordBatch`, what is the best way to handle it?

like this([sql/limit.rs](https://github.com/apache/arrow-datafusion/blob/b6252771248da117dd3a5976d134a66a4bb60431/datafusion/core/tests/sql/limit.rs#L167)):
```rust
let table = MemTable::try_new(schema, vec![partition])?;
ctx.register_table("a", Arc::new(table))?;
```
or ([sql/aggregates.rs](https://github.com/apache/arrow-datafusion/blob/b6252771248da117dd3a5976d134a66a4bb60431/datafusion/core/tests/sql/aggregates.rs#L2064))
```rust
let provider = MemTable::try_new(Arc::new(schema), vec![vec![batch1], vec![batch2]])?;
ctx.register_table("t", Arc::new(provider))?;
```

**Comment by alamb** [comment]
> Some of the test code is not a single RecordBatch, what is the best way to handle it?

I think leaving these more advanced instances using `MemTable` is best 👍

File: datafusion-examples/examples/dataframe_in_memory.rs [hunk 1]
```diff
@@ -47,8 +46,7 @@ async fn main() -> Result<()> {
     let ctx = SessionContext::new();
 
     // declare a table in memory. In spark API, this corresponds to createDataFrame(...).
-    let provider = MemTable::try_new(schema, vec![vec![batch]])?;
-    ctx.register_table("t", Arc::new(provider))?;
+    ctx.register_batch("t", batch)?;
```
Comment (alamb) [inline datafusion-examples/examples/dataframe_in_memory.rs:49]: ❤️

File: datafusion/core/src/dataframe.rs [hunk 2]
```diff
@@ -1361,10 +1361,8 @@ mod tests {
             ],
         )?;
 
-        let table = crate::datasource::MemTable::try_new(schema, vec![vec![data]])?;
-
         let ctx = SessionContext::new();
-        ctx.register_table("test", Arc::new(table))?;
+        ctx.register_batch("test", data)?;
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:1365]: its so much nicer

File: datafusion/core/src/execution/context.rs [hunk 3]
```diff
@@ -230,6 +231,16 @@ impl SessionContext {
         self.table_factories.insert(file_type.to_string(), factory);
     }
 
+    /// Registers the RecordBatch as the specified table name
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:234]:
```suggestion
    /// Registers the [`RecordBatch`] as the specified table name
```

File: datafusion/core/src/execution/context.rs [hunk 4]
```diff
@@ -713,6 +724,20 @@ impl SessionContext {
         )))
     }
 
+    /// Creates a DataFrame for reading a custom RecordBatch
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:727]:
```suggestion
    /// Creates a [`DataFrame`] for reading a single [`RecordBatch`]
```

**Review by alamb — APPROVED** [review]
Thank you @BaymaxHWY  -- this is a really nice addition and removes one of the rough edges for using DataFusion 👍

**Pull request merged**